File: python/railroaded/models/route.py
from enum import Enum
from typing import Optional

import seared as s

from .stop_continuity import StopContinuity


class TransitType(Enum):
    '''
    An `Enum` describing the transit type of a route.
    '''
    LIGHT_RAIL = 0
    SUBWAY = 1
    RAIL = 2
    BUS = 3
    FERRY = 4
    CABLE_TRAM = 5
    CABLE_CAR = 6
    FUNICULAR = 7
    TROLLEYBUS = 11
    MONORAIL = 12


@s.seared
class Route(s.Seared):
    '''
    A GTFS dataclass model for records found in `routes.txt`. Identifies a 
    transit route.
    
    Attributes:
        id (str):
            the unique ID of the route
        agency_id (str):
            the unique ID of the agency the route belongs to
        color (str):
            the color associated with the route
        desc (Optional[str]):
            a description of the route
        dropoffs (StopContinuity):
            the continuity of dropoffs along the route
        long_name (Optional[str]):
            the full name of the route
        name (str):
            the name of the route
        network_id (Optional[str]):
            the unique ID of the network the route belongs to
        pickups (StopContinuity):
            the continuity of pickups along the route
        short_name (Optional[str]):
            the short name of the route
        sort_idx (int):
            the sort position index of the route
        text_color (str):
            the color to use for text drawn against `Route.color`
        type (TransitType):
            the `TransitType` of the route
        url (Optional[str]):
            the URL of a webpage about the route
    '''

    ### ATTRIBUTES ###
    # Model ID
    id: str = s.Str(data_key='route_id', required=True)
    '''the unique ID of the route'''
    
    # Foreign IDs
    agency_id: str = s.Str(missing='')
    '''the unique ID of the agency the route belongs to'''
    network_id: Optional[str] = s.Str()
    '''the unique ID of the network the route belongs to'''

    # Required fields
    color: str = s.Str(data_key='route_color', missing='FFFFFF', required=True)
    '''the color associated with the route'''
    dropoffs: StopContinuity = s.Enum(
        enum=StopContinuity, 
        missing=StopContinuity.NONE
    )
    '''the continuity of dropoffs along the route'''
    pickups: StopContinuity = s.Enum(
        enum=StopContinuity, 
        missing=StopContinuity.NONE
    )
    '''the continuity of pickups along the route'''
    sort_idx: int = s.Int(data_key='route_sort_order', missing=0)
    '''the sort position index of the route'''
    text_color: Optional[str] = s.Str(
        data_key='route_text_color', missing='000000'
    )
    '''the color to use for text drawn against `Route.color`'''
    type: TransitType = s.Enum(
        data_key='route_type', enum=TransitType, required=True
    )
    '''the `TransitType` of the route'''

    # Optional fields
    desc: Optional[str] = s.Str(data_key='route_desc')
    '''a description of the route'''
    long_name: Optional[str] = s.Str(data_key='route_long_name')
    '''the full name of the route'''
    short_name: Optional[str] = s.Str(data_key='route_short_name')
    '''the short name of the route'''
    url: Optional[str] = s.Str(data_key='route_url')
    '''the URL of a web page about the route'''


    ### PROPERTIES ###
    @property
    def name (self) -> str:
        '''the name of the route'''
        return self.short_name if self.long_name is None else self.long_name

File: typescript/src/models/route.ts
import type { StopContinuity, TransitType } from "../types"
import { 
    fromStopContinuity, 
    fromTransitType, 
    toStopContinuity, 
    toTransitType 
} from "../types"
import { parseOptionalInt } from '../util'


export type GTFSRoute = {
    route_id: string
    agency_id?: string
    network_id?: string
    dropoffs?: string
    pickups?: string
    route_color?: string
    route_desc?: string
    route_long_name?: string
    route_short_name?: string
    route_type: string
    route_url?: string
    sort_idx?: string
    text_color?: string
}


export type MGTFSRoute = {
    id: string
    agency_id: string
    network_id?: string
    color: string
    desc?: string
    dropoffs: number
    long_name?: string
    pickups: number
    short_name?: string
    sort_idx: number
    text_color: string
    type: number
    url?: string
}


/**
 * A GTFS model for records found in `routes.txt`. Identifies a transit route.
 * 
 * @property id - the unique ID of the route
 * @property agencyId - the unique ID of the agency the route belongs to
 * @property networkId - the unique ID of the network the route belongs to
 * @property color - the color associated with the route
 * @property desc - a description of the route
 * @property dropoffs - the continuity of dropoffs along the route
 * @property longName - the full name of the route
 * @property name - the name of the route
 * @property pickups - the continuity of pickups along the route
 * @property shortName - the short name of the route
 * @property sortIdx - the sort index of the route
 * @property textColor - the color to use for text drawn against `color`
 * @property type - the `TransitType` of the route
 * @property url - the URL of a webpage about the route
 */
export default class Route {

    id: string
    agencyId: string
    networkId?: string
    color: string
    desc?: string
    dropoffs: StopContinuity
    longName?: string
    name: string
    pickups: StopContinuity
    shortName?: string
    sortIdx: number
    textColor: string
    type: TransitType
    url?: string

    constructor (data: MGTFSRoute) {
        this.id = data.id
        this.agencyId = data.agency_id
        this.networkId = data.network_id
        this.color = data.color
        this.desc = data.desc
        this.dropoffs = toStopContinuity(data.dropoffs)
        this.longName = data.long_name
        this.name = data.long_name ?? data.short_name ?? ''
        this.pickups = toStopContinuity(data.pickups)
        this.shortName = data.short_name
        this.sortIdx = data.sort_idx
        this.textColor = data.text_color
        this.type = toTransitType(data.type)
        this.url = data.url
    }

    static fromGTFS (data: GTFSRoute): MGTFSRoute {
        return {
            id: data.route_id,
            agency_id: data.agency_id ?? '',
            network_id: data.network_id,
            color: data.route_color ?? 'FFFFFF',
            desc: data.route_desc,
            dropoffs: parseOptionalInt(
                data.dropoffs,
                fromStopContinuity('none')
            ),
            long_name: data.route_long_name,
            pickups: parseOptionalInt(
                data.pickups,
                fromStopContinuity('none')
            ),
            short_name: data.route_short_name,
            sort_idx: parseOptionalInt(data.sort_idx, 0),
            text_color: data.text_color ?? '000000',
            type: parseInt(data.route_type),
            url: data.route_url
        }
    }

    toMGTFS (): MGTFSRoute {
        return {
            id: this.id,
            agency_id: this.agencyId,
            network_id: this.networkId,
            color: this.color,
            desc: this.desc,
            dropoffs: fromStopContinuity(this.dropoffs),
            long_name: this.longName,
            pickups: fromStopContinuity(this.pickups),
            short_name: this.shortName,
            sort_idx: this.sortIdx,
            text_color: this.textColor,
            type: fromTransitType(this.type),
            url: this.url
        }
    }

}